internal/serve/auth/signature_verifier.go

```diff
@@ -0,0 +1,157 @@
+	"strings"
+	"time"
+
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/strkey"
+	"github.com/stellar/go/support/log"
+)
+
+type SignatureVerifier interface {
+	VerifySignature(ctx context.Context, signatureHeaderContent string, rawReqBody []byte) error
+}
+
+var (
+	ErrStellarSignatureNotVerified = errors.New("neither Signature nor X-Stellar-Signature header could be verified")
+)
+
+type ErrInvalidTimestampFormat struct {
+	TimestampString string
+}
+
+func (e ErrInvalidTimestampFormat) Error() string {
+	return fmt.Sprintf("signature format different than expected. expected unix seconds, got: %s", e.TimestampString)
+}
+
+type ErrExpiredSignatureTimestamp struct {
+	ExpiredSignatureTimestamp time.Time
+	CheckTime                 time.Time
+}
+
+func (e ErrExpiredSignatureTimestamp) Error() string {
+	return fmt.Sprintf("signature timestamp has expired. sig timestamp: %s, check time %s", e.ExpiredSignatureTimestamp.Format(time.RFC3339), e.CheckTime.Format(time.RFC3339))
+}
+
+type StellarSignatureVerifier struct {
+	ServerHostname   string
+	WalletSigningKey string
+}
+
+var _ SignatureVerifier = (*StellarSignatureVerifier)(nil)
+
+// VerifySignature verifies the Signature or X-Stellar-Signature content and checks if the signature is signed for a known caller.
+func (sv *StellarSignatureVerifier) VerifySignature(ctx context.Context, signatureHeaderContent string, rawReqBody []byte) error {
+	t, s, err := ExtractTimestampedSignature(signatureHeaderContent)
+	if err != nil {
+		log.Ctx(ctx).Error(err)
+		var errTimestampFormat *ErrInvalidTimestampFormat
+		if errors.As(err, &errTimestampFormat) {
+			return err
+		}
+		return ErrStellarSignatureNotVerified
+	}
+
+	// 2 minutes in seconds
+	err = VerifyGracePeriodSeconds(t, 2*time.Second)
+	if err != nil {
+		log.Ctx(ctx).Error(err)
+		var errExpiredSignatureTimestamp *ErrExpiredSignatureTimestamp
+		if errors.As(err, &errExpiredSignatureTimestamp) {
+			return err
+		}
+		return ErrStellarSignatureNotVerified
+	}
+
+	signatureBytes, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		log.Ctx(ctx).Errorf("unable to decode signature value %s: %s", s, err.Error())
+		return ErrStellarSignatureNotVerified
+	}
+
+	payload := t + "." + sv.ServerHostname + "." + string(rawReqBody)
+
+	// TODO: perhaps add possibility to have more than one signing key.
+	kp, err := keypair.ParseAddress(sv.WalletSigningKey)
+	if err != nil {
+		return fmt.Errorf("parsing wallet signing key %s: %w", sv.WalletSigningKey, err)
+	}
+
+	err = kp.Verify([]byte(payload), signatureBytes)
+	if err != nil {
+		log.Ctx(ctx).Errorf("unable to verify the signature: %s", err.Error())
+		return ErrStellarSignatureNotVerified
+	}
+
+	return nil
+}
+
+func ExtractTimestampedSignature(signatureHeaderContent string) (t string, s string, err error) {
+	parts := strings.SplitN(signatureHeaderContent, ",", 2)
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("malformed header format: %s", signatureHeaderContent)
```

Comment:
Very nit-picky but I think we could lose the "format" at the end, since "malformed" already gives away that the format is wrong.
```suggestion
		return "", "", fmt.Errorf("malformed header: %s", signatureHeaderContent)
```